### PR TITLE
feature/qrcodegeneration

### PR DESCRIPTION
**Pull request by Matathiasq13824**
Added a generator of QR code.

### CODE REVIEW

**Review by Voxanimus — COMMENTED** [review]
Le qrCode est généré à partir de quel type de donnée ?
J'ai l'impression que y a des attribut de texte. et je suis pas bien sûr de comprendre pour quoi.

**Review by Voxanimus — APPROVED** [review]
tout me semble okay, ma question sur le texte sera réglable plus tard je pense

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import android.view.View
+import android.widget.ImageView
+import androidx.test.espresso.Espresso
+import androidx.test.espresso.action.ViewActions
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.intent.Intents
+import androidx.test.espresso.intent.matcher.IntentMatchers
+import androidx.test.espresso.matcher.BoundedMatcher
+import androidx.test.espresso.matcher.ViewMatchers
+import androidx.test.espresso.matcher.ViewMatchers.assertThat
```
Comment (felko) [inline app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt:15]: Looks like this import is not used if I believe Android Studio.

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+import androidx.test.ext.junit.rules.ActivityScenarioRule
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.google.zxing.*
+import com.google.zxing.common.HybridBinarizer
+import com.google.zxing.qrcode.QRCodeReader
+import org.hamcrest.CoreMatchers.equalTo
```
Comment (felko) [inline app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt:21]: Same here

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+//import org.robolectric.shadows.ShadowToast
```
Comment (felko) [inline app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt:28]: I think you can remove this until we get get roboelectric to work on the CI

File: app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            }
+
+            private fun extractContent(item: ImageView): String {
+                val drawable: Drawable = item.drawable
+                require(drawable is BitmapDrawable) { "The provided ImageView does not contain a bitmap" }
+                val binary = convertToBinary(drawable.bitmap)
+                return try {
+                    // Parse the bitmap and check it against expected value
+                    QRCodeReader().decode(binary).text
+                } catch (e: NotFoundException) {
```
Comment (felko) [inline app/src/androidTest/java/com/github/multimatum_team/multimatum/QRGeneratorTest.kt:105]: Slight nitpick:
It's sad that you can't catch multiple exception types at once but I think you can deduplicate the error message by doing:
```kotlin
try {
    ...
} catch (ex: Exception) {
    when(ex) {
        is NotFoundException, is ChecksumException, is FormatException -> {
            throw IllegalArgumentException("The provided image is not a valid QRCode", e)
        }
        else -> throw ex
    }
}
```
(see <https://stackoverflow.com/a/36760590/5559663>)

File: app/src/main/java/com/github/multimatum_team/multimatum/QRGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import androidx.appcompat.app.AppCompatActivity
+import android.os.Bundle
+import android.view.View
+import android.widget.EditText
+import android.widget.ImageView
+import android.widget.Toast
+import com.google.zxing.BarcodeFormat
+import com.google.zxing.qrcode.QRCodeWriter
+
+class QRGenerator : AppCompatActivity() {
```
Comment (felko) [inline app/src/main/java/com/github/multimatum_team/multimatum/QRGenerator.kt:15]: Would be nice to have a small docstring explaining what the activity does and briefly describe the interface

**Review by felko — CHANGES_REQUESTED** [review]

**Review by felko — APPROVED** [review]

**Pull request merged**